File: master-node-docker/sentinel/node/alive.py
# coding=utf-8
import time
from _thread import start_new_thread

from ..db import db


class UpdateNodesStatus(object):
    def __init__(self, max_secs):
        self.max_secs = max_secs
        self.stop_thread = False
        self.t = None

    def update_thread(self):
        while True:
            if self.stop_thread:
                break
            min_time = int(time.time()) - self.max_secs
            _ = db.nodes.update_many({
                'vpn.last_ping': {
                    '$lt': min_time
                }
            }, {
                '$set': {
                    'vpn.status': 'down'
                }
            })
            time.sleep(5)

    def start(self):
        self.t = start_new_thread(self.update_thread, ())

    def stop(self):
        self.stop_thread = True

File: master-node-nodejs/src/jobs/alive.js
import { scheduleJob } from "node-schedule";
import { waterfall } from "async";
import dbo from "../db/db";


export const alive = (data) => {
  let maxSecs = data.maxSecs || null;
  let db = null

  if (data.message == 'start') {
    scheduleJob('*/5 * * * * *', function () {
      let minTime = Date.now() / 1000 - maxSecs;
      if (global.db) {
        waterfall([
          (next) => {
            db = global.db
            next()
          }, (next) => {
            db.collection('nodes').updateMany({
              'vpn.ping_on': {
                '$lt': minTime
              }
            }, {
                '$set': {
                  'vpn.status': 'down'
                }
              }, (err, resp) => {
                if (err) throw err
                else next()
              })
          }
        ], (err, resp) => {
          console.log('alive')
        })
      }
    })
  }
}